File: packages/shared/src/protocol.ts
/**
 * Panel ↔ Node-Agent WebSocket protocol.
 *
 * The Go node-agent (`apps/node-agent/internal/ws`) and the panel's console
 * gateway (`apps/panel-api/src/agent/console.gateway.ts`) speak this exact
 * envelope. Browser clients connect to the panel gateway, which relays to the
 * agent; the message types are identical end-to-end.
 *
 * Every frame is JSON: `{ "type": <MessageType>, "payload": <object> }`.
 */

import type { ServerState } from './enums.js';

export const MessageType = {
  // Auth (first frame from client → gateway/agent)
  AUTH: 'auth',
  AUTH_OK: 'auth.ok',
  AUTH_ERROR: 'auth.error',

  // Console
  CONSOLE_OUTPUT: 'console.output', // agent → client (a line of stdout/stderr)
  CONSOLE_COMMAND: 'console.command', // client → agent (a line written to stdin)

  // Install / reinstall / game-switch streams
  INSTALL_OUTPUT: 'install.output',

  // Power
  POWER_COMMAND: 'power.command', // client → agent: { action: PowerAction }
  POWER_EVENT: 'power.event', // agent → client: { state: ServerState }

  // Live stats
  STATS_SUBSCRIBE: 'stats.subscribe', // client → agent
  STATS: 'stats', // agent → client (ResourceStats)
} as const;
export type MessageType = (typeof MessageType)[keyof typeof MessageType];

export const PowerAction = {
  START: 'start',
  STOP: 'stop',
  RESTART: 'restart',
  KILL: 'kill',
} as const;
export type PowerAction = (typeof PowerAction)[keyof typeof PowerAction];

export interface WsEnvelope<T = unknown> {
  type: MessageType;
  payload: T;
}

export interface AuthPayload {
  /** Short-lived ticket issued by the panel for this server/connection. */
  ticket: string;
  serverId: string;
}

export interface ConsoleOutputPayload {
  line: string;
  stream: 'stdout' | 'stderr';
  ts: number;
}

export interface ConsoleCommandPayload {
  command: string;
}

export interface PowerCommandPayload {
  action: PowerAction;
}

export interface PowerEventPayload {
  state: ServerState;
}

export interface ResourceStats {
  cpuPct: number;
  memUsedMb: number;
  memLimitMb: number;
  diskUsedMb: number;
  netRxBytes: number;
  netTxBytes: number;
  players?: number;
  uptimeSec?: number;
}

export type StatsPayload = ResourceStats;

/** Helper to build a well-typed frame. */
export function frame<T>(type: MessageType, payload: T): WsEnvelope<T> {
  return { type, payload };
}

File: apps/node-agent/internal/ws/protocol.go
// Package ws implements the WebSocket hub that relays live console output and
// stats to subscribed panel/browser clients and accepts console input from them.
package ws

import "encoding/json"

// MessageType enumerates the agent's WebSocket protocol message types. The wire
// format is a small JSON envelope {type, payload} in both directions.
type MessageType string

const (
	// Server -> client.
	TypeConsoleOutput MessageType = "console.output" // payload: {line}
	TypeStats         MessageType = "stats"          // payload: stats sample
	TypePowerEvent    MessageType = "power.event"    // payload: {state}
	TypeInstallOutput MessageType = "install.output" // payload: {line, done}
	TypeError         MessageType = "error"          // payload: {message}
	TypeAuthOK        MessageType = "auth.ok"

	// Client -> server.
	TypeAuth           MessageType = "auth"            // payload: {token}
	TypeConsoleCommand MessageType = "console.command" // payload: {command}
	TypePowerCommand   MessageType = "power.command"   // payload: {action}
	TypeSubscribeStats MessageType = "stats.subscribe"
)

// Message is the protocol envelope.
type Message struct {
	Type    MessageType     `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// NewMessage builds a Message with a JSON-encoded payload.
func NewMessage(t MessageType, payload any) (Message, error) {
	b, err := json.Marshal(payload)
	if err != nil {
		return Message{}, err
	}
	return Message{Type: t, Payload: b}, nil
}

// Payload helpers for the common shapes.

type ConsoleLine struct {
	Line string `json:"line"`
}

type CommandPayload struct {
	Command string `json:"command"`
}

type PowerPayload struct {
	Action string `json:"action"` // start|stop|restart|kill
	State  string `json:"state,omitempty"`
}

type AuthPayload struct {
	Token string `json:"token"`
}

type ErrorPayload struct {
	Message string `json:"message"`
}
